python/cudf/bindings/join.pyx

```diff
@@ -78,7 +77,7 @@ cpdef join(col_lhs, col_rhs, left_on, right_on, how, method='sort'):
         if col.has_null_mask:
             mask_size = col.nullmask.size
 
-        if name not in right_on:
+        if name not in on:
             result_cols.push_back(column_view_from_NDArrays(0, None, mask=None, dtype=col._column.dtype, null_count=0))
             result_col_names.append(name)
```

Comment:
You've introduced some redundancy here - the two blocks above `if name not in on:` are identical - previously one pushed `left_on` columns first, then left columns from the overlap of left and right on, then right columns. In the new refactor it'll first push the non-on columns, then the on columns, then the non-on columns again.

Similarly but more straightforward, `for name in on: if name in on:` is more redundant.

Maybe there's logic here that I am missing - I'm responsible for a lot of this particular `merge` structuring, and I couldn't figure out in the time I had how to do `left_on, right_on` correctly, so maybe I'm missing something. Good job for improving this!

Reply:
Yeah, I have messed up the `right_on`, `left_on`, and `on` argument. 
I have change this `join()` function to only take `right_on` and `left_on`.